backend-go/lib.go

```diff
@@ -0,0 +1,324 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"github.com/mitchellh/mapstructure"
+	"github.com/nats-io/nats.go"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+var Config TelestionBaseConfig
+
+type TelestionError struct {
+	Error   *error // only present if the TelestionError is based on an error
+	Message string
+	Code    int // might be interesting later if we want to identify where the error is coming from
+}
+
+func (t *TelestionError) FormatError() string {
+	return fmt.Sprint(
+		"An unexpected error (code: ",
+		t.Code,
+		") occurred in the Telestion backend library: \"",
+		t.Message,
+		"\"")
+}
+
+func NewTelestionError(err error, msg string) *TelestionError {
+	return &TelestionError{&err, msg, 1}
+}
+
+// Config parsing process must at least yield the following minimal config scheme
+type minimalConfig struct {
+	Nats        bool   `json:"NATS"`
+	NatsUrl     string `json:"NATS_URL"`
+	ServiceName string `json:"SERVICE_NAME"`
+	DataDir     string `json:"DATA_DIR"`
+}
+
+func inferMinimalConfig(mapping map[string]any) (*minimalConfig, *TelestionError) {
+	mConf := minimalConfig{}
+
+	decoderConfig := &mapstructure.DecoderConfig{
+		ErrorUnused: false,
+		ErrorUnset:  true,
+		Result:      &mConf,
+	}
+	mDecoder, err := mapstructure.NewDecoder(decoderConfig)
+	if err != nil {
+		return nil, NewTelestionError(err, "Decoder for minimal config inference could not be initialized!")
+	}
+
+	if err := mDecoder.Decode(mapping); err != nil {
+		return nil, NewTelestionError(err, "Minimal config could not be inferred from given map!")
```

Comment:
This prints out:

```text
Error starting Telestion service: An unexpected error (code: 1) occurred in the Telestion backend library: "Minimal config could not be inferred from given map!"
```

However, this isn't really true – it's not unexpected that an incomplete config gets passed. Instead, the message should be actionable and specify what the user can do to resolve this (in this case, pass the necessary config).

Reply:
No longer applicable because we do longer use a custom error type. :wink: